### PR TITLE
Techdebt/provider refactor + blocknative wip

### PR DESCRIPTION
**Pull request by jmzwar**
This PR upgrades our blocknative bnc-onboard wallet to web3-onboard, as well as refactoring much of the global state management to centre around events rather than individual pieces of state.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -42,6 +42,15 @@
...
+		"@web3-onboard/gnosis": "^2.0.6",
+		"@web3-onboard/injected-wallets": "^2.0.12",
+		"@web3-onboard/ledger": "^2.1.5",
+		"@web3-onboard/portis": "^2.0.5",
+		"@web3-onboard/torus": "^2.0.6",
+		"@web3-onboard/trezor": "^2.1.5",
+		"@web3-onboard/walletconnect": "^2.0.6",
 		"axios": "^0.21.1",
 		"bignumber.js": "^9.0.0",
 		"bnc-onboard": "^1.36.0",
```
Comment (bachstatter) [inline package.json:56]: We can remove this now, right?

**Review by bachstatter — APPROVED** [review]
Nice! I left some minor comments but it looks good to me.
We still have the problem that staking will ask metamask to swtich network if you have our tab open and are on a network we're not supporting. But if you'd plan to tackle that in another PR this make sense.
Oh also, need to rebase dev too, have some conflicts

File: components/WalletComponents/index.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+</svg>
+`;
```
[the remark this answers is not in the document]
Comment (jmzwar) [reply components/WalletComponents/index.tsx:4]: Blocknative only accepts a string for this.

File: containers/Connector/Connector.tsx [hunk 3]
```diff
@@ -1,72 +1,131 @@
...
-	});
-	const [signer, setSigner] = useState<ethers.Signer | null>(null);
-	const [synthetixjs, setSynthetixjs] = useState<SynthetixJS | null>(null);
-	const [onboard, setOnboard] = useState<ReturnType<typeof initOnboard> | null>(null);
-	const [isAppReady, setAppReady] = useRecoilState(appReadyState);
-	const [walletAddress, setWalletAddress] = useRecoilState(walletAddressState);
-	const [selectedWallet, setSelectedWallet] = useLocalStorage<string | null>(
-		LOCAL_STORAGE_KEYS.SELECTED_WALLET,
-		''
+	const [state, dispatch] = useReducer(reducer, initialState);
```
[the remark this answers is not in the document]
Comment (jmzwar) [reply containers/Connector/Connector.tsx:29]: If you want shared state you can only use the hook in one place, and then use context/unstated to consume.

File: containers/Connector/reducer.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	| { type: 'app_ready'; payload: OnboardAPI }
+	| { type: 'config_update'; payload: ConnectionUpdate }
+	| { type: 'watch_wallet'; payload: WatchWallet }
+	| { type: 'set_ens'; payload: EnsUpdate }
+	| { type: 'update_provider'; payload: ProviderUpdate }
+	| { type: 'wallet_disconnected' };
+
+export function reducer(state: ConnectorState, action: Actions) {
+	switch (action.type) {
+		case 'app_ready':
```
[the remark this answers is not in the document]
Comment (jmzwar) [reply containers/Connector/reducer.ts:70]: Yeah I think I might actually change this to an UPPER_CASE enum.

File: package.json (hunk 1 above)
Comment (jmzwar) [reply package.json:56]: Yup of course!

File: containers/Connector/Connector.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmzwar) [reply containers/Connector/Connector.tsx:29]: That isn't how it works. This connector defined a context that has state. The context (including the state) is only defined once, it is then consumed in all those different places, not re-initialised.

File: containers/Connector/reducer.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+				...state,
+				network: null,
+				provider: null,
+				signer: null,
+				synthetixjs: null,
+				walletAddress: null,
+				watchedWallet: null,
+				ensName: null,
+				ensAvatar: null,
+			};
```
[the remark this answers is not in the document]
Comment (jmzwar) [reply containers/Connector/reducer.ts:108]: I'll add a comment here, but I'd like to keep the onboard instance the same.

File: sections/shared/modals/WalletOptionsModal/WalletOptionsModal.tsx [hunk 6]
```diff
@@ -123,12 +79,10 @@ const WalletOptionsModal: FC<WalletOptionsProps> = ({
...
 				<>
 					<WalletDetails>
-						{walletWatched ? (
+						{walletWatched && (
 							<SelectedWallet>
 								<Incognito />
 							</SelectedWallet>
-						) : (
-							<SelectedWallet>{getWalletIcon(selectedWallet?.toLowerCase())}</SelectedWallet>
 						)}
```
[the remark this answers is not in the document]
Comment (jmzwar) [reply sections/shared/modals/WalletOptionsModal/WalletOptionsModal.tsx:86]: Hmm, let me have a closer look at this.

**Comment by jmzwar** [comment]
> Nice! I left some minor comments but it looks good to me. We still have the problem that staking will ask metamask to swtich network if you have our tab open and are on a network we're not supporting. But if you'd plan to tackle that in another PR this make sense. Oh also, need to rebase dev too, have some conflicts

@bachstatter I've added this for the network switching thing
Screen Shot 2022-07-20 at 1 32 55 pm

And re-added some icons for wallet type

**Review by bachstatter — APPROVED** [review]
Looks good! Nice work

**Pull request merged**